core/services.go

```diff
@@ -362,24 +365,211 @@ func (app *Application) uploadProfileImageToAws(image image.Image, filename stri
+	item.OrgID = claims.OrgID
+	item.DateCreated = time.Now().UTC()
+	item, err := s.app.storage.CreateDataContentItem(item)
+	if err != nil {
+		return nil, err
+	}
+	return item, nil
+}
+
+func (s *servicesImpl) UpdateDataContentItem(claims *tokenauth.Claims, item *model.DataContentItem) (*model.DataContentItem, error) {
+	var dataItem *model.DataContentItem
+
+	category, err := s.app.storage.FindCategory(&claims.AppID, claims.OrgID, item.Category)
+	if err != nil {
+		return nil, err
+	}
+
+	isValid := false
+	for _, element := range category.Permissions {
+		if strings.Contains(claims.Permissions, element) {
+			isValid = true
+			break
+		}
+	}
+
+	//TODO check new permission
+
+	if !isValid {
+		return nil, fmt.Errorf("Unauthorized to update data content item")
+	}
+
+	dataItem, err = s.app.storage.UpdateDataContentItem(&claims.AppID, claims.OrgID, item)
+	if err != nil {
+		return nil, err
+	}
+
+	return dataItem, err
+}
+
+func (s *servicesImpl) DeleteDataContentItem(claims *tokenauth.Claims, key string) error {
+
+	item, err := s.app.storage.FindDataContentItem(&claims.AppID, claims.OrgID, key)
+	if err != nil {
+		return err
+	}
+
+	category, err := s.app.storage.FindCategory(&claims.AppID, claims.OrgID, item.Category)
+	if err != nil {
+		return err
+	}
+
+	isValid := false
+	for _, element := range category.Permissions {
+		if strings.Contains(claims.Permissions, element) {
+			isValid = true
+			break
+		}
+	}
+
+	if !isValid {
+		return fmt.Errorf("Unauthorized to delete data content item")
+	}
+
+	err = s.app.storage.DeleteDataContentItem(&claims.AppID, claims.OrgID, key)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *servicesImpl) CreateCategory(claims *tokenauth.Claims, item *model.Category) (*model.Category, error) {
+	item.ID = uuid.NewString()
+	item.AppID = &claims.AppID
+	item.OrgID = claims.OrgID
+	item.DateCreated = time.Now().UTC()
+	item, err := s.app.storage.CreateCategory(item)
+	if err != nil {
+		return nil, err
+	}
+	return item, nil
+}
+
+func (s *servicesImpl) GetCategory(appID string, orgID string, id string) (*model.Category, error) {
+	item, err := s.app.storage.FindCategory(&appID, orgID, id)
+	if err != nil {
+		return nil, err
+	}
+	return item, nil
+}
+
+func (s *servicesImpl) UpdateCategory(appID string, orgID string, item *model.Category) (*model.Category, error) {
+	item, err := s.app.storage.UpdateCategory(&appID, orgID, item)
+	if err != nil {
+		return nil, err
+	}
+	return item, nil
+}
+
+func (s *servicesImpl) DeleteCategory(appID string, orgID string, id string) error {
+	err := s.app.storage.DeleteCategory(&appID, orgID, id)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *servicesImpl) UploadFileContentItem(file io.Reader, claims *tokenauth.Claims, fileName string, category string) error {
+
+	path := claims.OrgID + "/" + claims.AppID + "/" + category + "/user/" + claims.Subject + "/" + fileName
+
+	_, err := s.app.awsAdapter.UploadFile(file, path)
```

Comment:
Let's check the category permissions against the token permissions before uploading.